### PR TITLE
Optimize ReorderGlobals ordering

### PR DESCRIPTION
**Pull request by kripken**
The old ordering in that pass did a topological sort while sorting by uses
both within topological groups and between them. That could be unoptimal
in some cases, however, and actually on J2CL output this pass made the
binary larger, which is how we noticed this.

The problem is that such a toplogical sort keeps topological groups in
place, but it can be useful to interleave them sometimes. Imagine this:
```
     $c - $a
    /
  $e
    \
     $d - $b
```
Here `$e` depends on `$c`, etc. The optimal order may interleave the two
arms here, e.g. `$a, $b, $d, $c, $e`. That is because the dependencies define
a partial order, and so the arms here are actually independent.

Sorting by toplogical depth first might help in some cases, but also is not
optimal in general, as we may want to mix toplogical depths:
`$a, $c, $b, $d, $e` does so, and it may be the best ordering.

This PR implements a natural greedy algorithm that picks the global with
the highest use count at each step, out of the set of possible globals, which
is the set of globals that have no unresolved dependencies. So we start by
picking the first global with no dependencies and add at at the front; then
that unlocks anything that depended on it and we pick from that set, and
so forth.

This may also not be optimal, but it is easy to make it more flexible by
customizing the counts, and we consider 4 sorts here:

1. Set all counts to 0. This means we only take into account dependencies,
  and we break ties by the original order, so this is as close to the original
  order as we can be.
2. Use the actual use counts. This is the simple greedy algorithm.
3. Set the count of each global to also contain the counts of its children,
  so the count is the total that might be unlocked. This gives more weight
  to globals that can unlock more later, so it is less greedy.
4. As 3, but weight children's counts lower in an exponential way, which
  makes sense as they may depend on other globals too.

In practice it is simple to generate cases where 1, 2, or 3 is optimal (see
new tests), but on real-world J2CL I see that 4 (with a particular exponential
coefficient) is best, so the pass computes all 4 and picks the best. As a
result it will never worsen the size and it has a good chance of
improving.

The differences between these are small, so in theory we could pick any
of them, but given they are all modifications of a single algorithm it is
very easy to compute them all with little code complexity.

Some data on J2CL:

| Method | Size (less is better) |
|---|---|
| Greedy | 4338066 |
| Original order | 4336535 |
| Old (before this PR) |4336518 |
| Sum | 4336503 |
| Exponential (0.5) | 4336373 |
| Exponential (PR's value) | 4336127 |

There is a slight runtime cost to this: J2CL goes from 0.9666 to 1.1351
seconds. As this is one of our faster passes the slight slowdown seems
worth it in return for the guarantee to never increase size, and the small
improvement.

### CODE REVIEW

**Comment by gkdn** [comment]
Interesting problem! (cc @rluble)

The number differences here are small; makes me wonder if magic-import that pushes higher indices benefit more.

File: src/passes/ReorderGlobals.cpp [hunk 1]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+      // must be in bounds.
+      assert(indices[i] < indices.size());
+      assert(actualOrder[indices[i]] == 0);
+
+      actualOrder[indices[i]] = i;
+    }
+
+    if (always) {
+      // In this mode we gradually increase the cost of later globals, in an
+      // unrealistic but manner.
```
Comment (tlively) [inline src/passes/ReorderGlobals.cpp:383]: Missing word?

**Review by tlively — COMMENTED** [review]
Was there a test for which the exponential sort was optimal?

**Comment by kripken** [comment]
> Was there a test for which the exponential sort was optimal?

Sadly no. It's very hard to make such a test, as it would need very many deep dependency chains (on which it pays to take them into account a little, but not too much).

(I would not include code for that in the pass, except that it ends up as just another constant value, and we do measure the sizes, so it seems safe.)

File: src/passes/ReorderGlobals.cpp [hunk 2]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+    // them (that is, we add a small multiple of the dependent's counts, which
+    // may help as the dependents also depend on other things potentially, so a
+    // simple sum may be too naive).
+    double const EXPONENTIAL_FACTOR = 0.095;
+    IndexCountMap sumCounts(globals.size()), exponentialCounts(globals.size());
+    // We add items to |sumCounts, exponentialCounts| after they are computed,
+    // so anything not there must be pushed to a stack before we can handle it.
+    std::vector<Index> stack;
+    // Do a simple recursion to compute children before parents, while tracking
+    // which have been computed.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp]: Done. It's just a tiny bit shorter but I guess it is less bug-prone to reuse code...

File: src/passes/ReorderGlobals.cpp [hunk 3]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+    // Apply the indices we computed.
     std::sort(
-      module->globals.begin(),
-      module->globals.end(),
+      globals.begin(),
+      globals.end(),
       [&](const std::unique_ptr<Global>& a, const std::unique_ptr<Global>& b) {
-        return sortedIndexes[a->name] < sortedIndexes[b->name];
+        return (*best)[originalIndices[a->name]] <
+               (*best)[originalIndices[b->name]];
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp]: I did this now, good idea... it is a little more code, but it is actually 10% faster... :rocket:

File: src/passes/ReorderGlobals.cpp [hunk 4]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
     module->updateMaps();
   }
+
+  IndexIndexMap doSort(const IndexCountMap& counts,
+                       const Dependencies& originalDeps,
+                       Module* module) {
+    auto& globals = module->globals;
+
+    // Copy the deps as we will operate on them as we go.
+    auto deps = originalDeps;
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:253]: True. But the explicit copy with a comment seems clearer to me.

File: src/passes/ReorderGlobals.cpp [hunk 5]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+      // anyhow, but it is better to do it here in the IR so we can actually
+      // see what the final layout will be.
+      auto aImported = globals[a]->imported();
+      auto bImported = globals[b]->imported();
+      if (!aImported && bImported) {
+        return true;
+      }
+      if (aImported && !bImported) {
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:302]: Hmm, that is shorter, but it is harder to read, isn't it? That `bImported` is the outcome means I need to read that line very carefully...

File: src/passes/ReorderGlobals.cpp [hunk 6]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+    // higher use count than $a. For that reason we try several variations of
+    // this with different counts, see earlier.
+    std::vector<Index> availableHeap;
+
+    // Comparison function. This is used in a heap, where "highest" means
+    // "popped first", so if we compare the counts 10 and 20 we will return 1,
+    // which means 10 appears first in a simple sort, as 20 is higher. Later, 20
+    // will be popped earlier as it is higher and then it will appear earlier in
+    // the actual final sort.
+    auto cmp = [&](Index a, Index b) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:291]: > it looks like this implements greater than

The code has 
```cpp
      if (aCount < bCount) {
        return true;
      }
```
So it checks less than?

File: src/passes/ReorderGlobals.cpp [hunk 7]
```diff
@@ -53,112 +64,354 @@ struct UseCountScanner : public WalkerPass<PostWalker<UseCountScanner>> {
...
+      }
+    }
+
+    // Pop off the heap: Emit the global and it its final, sorted index. Keep
+    // doing that until we finish processing all the globals.
+    IndexIndexMap sortedindices(globals.size());
+    Index numSortedindices = 0;
+    while (!availableHeap.empty()) {
+      std::pop_heap(availableHeap.begin(), availableHeap.end(), cmp);
+      auto global = availableHeap.back();
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:335]: This is an `Index`, so it seems less clear (and maybe risky) to use a reference here? (risky as we'd need to move all uses of the reference to before the `pop_back` which invalidates it, which is not the case atm)

File: src/passes/ReorderGlobals.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/ReorderGlobals.cpp]: Oops, yeah, fixed.

File: src/passes/ReorderGlobals.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/ReorderGlobals.cpp:336]: Ok, I think I thought it was something more expensive than an index.

File: src/passes/ReorderGlobals.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/ReorderGlobals.cpp:292]: Oh hmm, I think it's the `return a > b` fallback that threw me off. Why isn't it `return a < b`?

File: src/passes/ReorderGlobals.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/ReorderGlobals.cpp:303]: Fair. I like it for symmetry with the next check, but don't feel strongly about it.

File: src/passes/ReorderGlobals.cpp [hunk 8]
```diff
@@ -237,14 +236,12 @@ struct ReorderGlobals : public Pass {
...
-    std::sort(
-      globals.begin(),
-      globals.end(),
-      [&](const std::unique_ptr<Global>& a, const std::unique_ptr<Global>& b) {
-        return (*best)[originalIndices[a->name]] <
-               (*best)[originalIndices[b->name]];
-      });
-
+    std::vector<std::unique_ptr<Global>> old;
+    old.swap(globals);
```
Comment (tlively) [inline src/passes/ReorderGlobals.cpp:240]:
```suggestion
    std::vector<std::unique_ptr<Global>> old = std::move(globals);
```

**Review by tlively — APPROVED** [review]

File: src/passes/ReorderGlobals.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:290]: These things do work in opposite orders. We want to sort in increasing order by the original order (so 0, 1, 2, 3 are the indexes in the final order), and decreasing order in the counts (so 100, 90, 80 are the use counts). Yeah, this is kind of confusing especially with the heap that pops off the top. I had a comment on this but I think it was more confusing than useful. I removed it now and added comments before each one of the comparisons. While doing that I found a way to use your suggestion for `return bImport` that shortens the code, with a comment now I think it is an improvement to what I had before.

File: src/passes/ReorderGlobals.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/ReorderGlobals.cpp]: Applied in the last commit.

File: src/passes/ReorderGlobals.cpp (hunk 8 above)
Comment (tlively) [reply src/passes/ReorderGlobals.cpp:240]: Hmm, it does appear that move assignment for vectors uses the "valid but unspecified" language. But their move constructor [further guarantees](https://en.cppreference.com/w/cpp/container/vector/vector) that the operand is left in the empty state.
```suggestion
    std::vector<std::unique_ptr<Global>> old(std::move(globals));
```

File: src/passes/ReorderGlobals.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/ReorderGlobals.cpp:290]: Great, thanks!

File: src/passes/ReorderGlobals.cpp (hunk 8 above)
Comment (kripken) [reply src/passes/ReorderGlobals.cpp:240]: Interesting. I guess specific cases are ok then, even if the general rule disallows it. But I think it's safer not to assume the specific case here, as we may replace the `std::vector` with some other data structure some day.

Comment (tlively) [reply src/passes/ReorderGlobals.cpp:240]: Right, `std::move` itself imposes no requirements except that the result be "valid." Individual types further specify what happens when they are moved. All container types should do the right thing here, so I think the extra code clarity is worth it.

Comment (kripken) [reply src/passes/ReorderGlobals.cpp:240]: Fair enough, I guess we would do the same in any custom data structures we use too.

**Pull request merged**